dev/src/dev/deps_graph.clj

```diff
@@ -56,3 +56,66 @@
   "All namespaces from a module that are used outside that module."
   [module-symb]
   (into (sorted-set) (map :depends-on-namespace) (external-usages module-symb)))
+
+(defn module-dependencies
+  "Build a graph of module => set of modules it refers to."
+  ([]
+   (letfn [(reduce-module-deps [module-deps module deps]
+             (reduce
+              (fn [module-deps {dep-module :module, :as _dep}]
+                (cond-> module-deps
+                  (not= dep-module module) (conj dep-module)))
+              (or module-deps (sorted-set))
+              deps))
+           (reduce-deps [module->deps {:keys [module deps]}]
+             (update module->deps module reduce-module-deps module deps))]
+     (reduce reduce-deps (sorted-map) (dependencies))))
+
+  ([module]
+   (get (module-dependencies) module)))
+
+(defn circular-dependencies
+  "Build a graph of module => set of modules it refers to that also refer to this module."
+  ([]
+   (let [module->deps (module-dependencies)]
+     (letfn [(circular-dependency? [module-x module-y]
+               (and (contains? (get module->deps module-x) module-y)
+                    (contains? (get module->deps module-y) module-x)))
+             (circular-deps [module]
+               (let [module-deps (get module->deps module)]
+                 (not-empty (into (sorted-set)
+                                  (filter (fn [dep]
+                                            (circular-dependency? module dep)))
+                                  module-deps))))]
+       (into (sorted-map)
+             (keep (fn [module]
+                     (when-let [circular-deps (circular-deps module)]
+                       [module circular-deps])))
+             (keys module->deps)))))
```

Comment:
Doesn't matter too much, but you might as well unify the 1-arity variant with the `circular-deps` helper as they have the same contract and basically the same name too. The overloaded version will avoid a bunch of necessary calculations, and the code will be a bit less nested too.

Haven't checked this refactor for dumb mistakes.

```suggestion
(defn circular-dependencies
  "Build a graph of module => set of modules it refers to that also refer to this module."
  ([]
   (let [module->deps (module-dependencies)]
     (into (sorted-map)
           (keep (fn [module]
                   (when-let [circular-deps (circular-dependencies module module->deps)]
                     [module circular-deps])))
           (keys module->deps))))
  ([module]
   (circular-dependencies module (module-dependencies)))
  ([module module->deps]
   (letfn [(circular-dependency? [module-x module-y]
             (and (contains? (get module->deps module-x) module-y)
                  (contains? (get module->deps module-y) module-x)))]
     (not-empty (into (sorted-set)
                      (filter (fn [dep]
                                (circular-dependency? module dep)))
                      (get module->deps module))))))
```